acceptance/tests/ssl/trusted_external_facts.rb

```diff
@@ -51,18 +56,110 @@ class external {
+      on(master, "chmod 777 #{external_trusted_fact_script_path}")
+    end
+
+    step "start the master and perform the test" do
+      master_opts = {
+        'main' => {
+          'trusted_external_command' => external_trusted_fact_script_path
+        }
       }
+
+      with_puppet_running_on(master, master_opts) do
+        agents.each do |agent|
+          on(agent, puppet("agent", "-t", "--environment", tmp_environment), :acceptable_exit_codes => [0,2]) do |res|
+            trusted_hash = parse_trusted_json(res.stdout)
+            assert_includes(trusted_hash, 'external', "Trusted fact hash contains external key")
+            assert_equal(agent.to_s, trusted_hash['external']['doot'], "trusted facts contains certname")
+          end
+        end
+      end
+    end
+  end
+
+  step "when trusted_external_command is a directory" do
+    dir_path = "#{testdir}/commands"
+    executable_files = {
+      'no_extension' => <<EOF,
+#!/bin/bash
+CERTNAME=$1
+printf '{"no_extension_key":"%s"}\n' "$CERTNAME"
+EOF
+
+      'shell.sh' => <<EOF,
+#!/bin/bash
+CERTNAME=$1
+printf '{"shell_key":"%s"}\n' "$CERTNAME"
+EOF
+
+      'ruby.rb' => <<EOF,
+#!#{master[:privatebindir]}/ruby
+require 'json'
+CERTNAME=ARGV[0]
+data = { "ruby_key" => CERTNAME }
+print data.to_json
+EOF
     }
 
-    with_puppet_running_on(master, master_opts) do
-      agents.each do |agent|
-        on(agent, puppet("agent", "-t", "--environment", tmp_environment), :acceptable_exit_codes => [0,2]) do |res|
-          assert_match(/external/, res.stdout, "Trusted fact hash contains external key")
-          assert_match(/doot.*#{agent}/, res.stdout, "trusted facts contains certname")
+    step "create the directory" do
+      on(master, "mkdir #{dir_path}")
+      on(master, "chmod 755 #{dir_path}")
+
+      executable_files.each do |filename, content|
+        filepath = "#{dir_path}/#{filename}"
+        create_remote_file(master, filepath, content)
+        on(master, "chmod 777 #{filepath}")
+      end
+
+      # Create a non-executable file and an executable child-directory
+      # to ensure that these cases are skipped during external data
+      # retrieval
+
+      create_remote_file(master, "#{dir_path}/non_executable_file", "foo")
+
+      executable_child_dir = "#{dir_path}/child_dir"
+      on(master, "mkdir #{executable_child_dir}")
+      on(master, "chmod 777 #{executable_child_dir}")
+    end
+
+    step "start the master and perform the test" do
+      master_opts = {
+        'main' => {
+          'trusted_external_command' => dir_path
+        }
+      }
+
+      with_puppet_running_on(master, master_opts) do
+        agents.each do |agent|
+          on(agent, puppet("agent", "-t", "--environment", tmp_environment), :acceptable_exit_codes => [0,2]) do |res|
+            trusted_hash = parse_trusted_json(res.stdout)
+            assert_includes(trusted_hash, 'external', "Trusted fact hash contains external key")
+
+
+            external_keys = [
```

Comment:
Could also calculate this from `executable_files` but meh, that's also kind of circular (using code we're testing to calculate expected result).